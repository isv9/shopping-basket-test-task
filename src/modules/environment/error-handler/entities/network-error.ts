export class NetworkError extends Error {
    constructor() {
        super('Network error');
        this.name = 'NetworkError';
    }
}
