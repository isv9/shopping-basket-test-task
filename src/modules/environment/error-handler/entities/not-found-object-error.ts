export class NotFoundObjectError extends Error {
    constructor(objectName: string) {
        super(`Not found '${objectName}'`);
        this.name = 'NotFoundObjectError';
    }
}
