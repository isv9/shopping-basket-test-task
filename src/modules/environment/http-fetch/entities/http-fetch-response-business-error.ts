export type HttpFetchResponseBusinessError = { error: string; type: string };