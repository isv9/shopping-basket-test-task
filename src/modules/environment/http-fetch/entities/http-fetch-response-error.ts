import {HttpFetchResponse} from "./http-fetch-response";

export interface HttpFetchResponseError<TErrorBody = any> extends Error {
    request?: any;
    response: HttpFetchResponse<TErrorBody>;
    code?: string;
}