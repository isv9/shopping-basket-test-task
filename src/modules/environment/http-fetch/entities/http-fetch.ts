import { HttpFetchResponse } from './http-fetch-response';

export interface HttpFetch {
    get<TResponse = object>(url: string): Promise<HttpFetchResponse<TResponse>>;

    post<TBody, TResponse = object>(url: string, data: TBody): Promise<HttpFetchResponse<TResponse>>;

    delete<TBody, TResponse = object>(url: string, data: TBody): Promise<HttpFetchResponse<TResponse>>;
}
