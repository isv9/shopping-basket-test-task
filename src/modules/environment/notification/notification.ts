export interface NotificationService {
    showError(errorMessage: string): void
}

export const notificationService: NotificationService = {
    showError: (errorMessage: string) => {
        alert(`Error: ${errorMessage}`);
    }
};