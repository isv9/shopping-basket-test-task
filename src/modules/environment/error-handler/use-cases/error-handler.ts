import {NotificationService} from "../../notification/notification";
import {NetworkError} from "../entities/network-error";

export class ErrorHandler {

    constructor(private readonly notificationService: NotificationService) {
        this.processError = this.processError.bind(this);
    }

    processError(error: Error | undefined) {
        if (error instanceof NetworkError) {
            this.notificationService.showError(error.message);
            return;
        }
        //and ...
        this.notificationService.showError('Unknown error');
    }
}
