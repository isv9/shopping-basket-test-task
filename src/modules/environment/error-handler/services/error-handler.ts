import {notificationService} from "../../notification/notification";
import {ErrorHandler} from "../use-cases/error-handler";

export const errorHandler = new ErrorHandler(notificationService);

