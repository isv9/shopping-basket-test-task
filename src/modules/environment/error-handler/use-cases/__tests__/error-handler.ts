import {ErrorHandler} from "../error-handler";
import {NotificationService} from "../../../notification/notification";
import {NetworkError} from "../../entities/network-error";

describe('ErrorHandler', () => {

    let errorHandler: ErrorHandler;
    let notificationService: NotificationService;

    beforeEach(() => {
        notificationService = {
            showError: jest.fn() as any,
        };
        errorHandler = new ErrorHandler(notificationService);
    });

    it('unknown error', () => {
        errorHandler.processError(new Error('test error'));
        expect((notificationService.showError as any).mock.calls).toMatchSnapshot();
    });

    it('NetworkError', () => {
        errorHandler.processError(new NetworkError());
        expect((notificationService.showError as any).mock.calls).toMatchSnapshot();
    });
});