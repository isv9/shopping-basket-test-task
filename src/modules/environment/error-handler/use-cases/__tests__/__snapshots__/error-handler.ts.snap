// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorHandler NetworkError 1`] = `
Array [
  Array [
    "Network error",
  ],
]
`;

exports[`ErrorHandler unknown error 1`] = `
Array [
  Array [
    "Unknown error",
  ],
]
`;
